[ONNX] Refactor ShapeInferenceWithFakeTensor to fill  metavalue into the original gm

titaiwangms opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98760

From https://github.com/pytorch/pytorch/pull/97494#discussion_r1160068456, the passes should modify gm inplace, but before this PR, `ShapeInferenceWithFakeTensor` utilizes Transform.transform() to make a copy of the gm, and rely on the assumption that the topological order of two graphs should be the same. This PR addresses the issue by saving another metavalue `static_shape` into gm for op_level_debug, instead of overwriting `val`.

cc @BowenBao

titaiwangms committed:
[ONNX] Refactor ShapeInferenceWithFakeTensor to fill  metavalue into the original gm

[ghstack-poisoned]

BowenBao commented on torch/onnx/_internal/fx/op_validation.py:
```diff
@@ -195,14 +195,13 @@ def _recursive_wrap_args(
-                # TODO(titaiwang): Could dtype be inside a fx.Node?
                 wrapped_args.append(fake_tensor)
             else:
                 warnings.warn(
```
Should we raise instead of warn?

titaiwangms replied:
I don't want op_level_debug to stop exporter. What do you think if we append unsupported types into torch_args, and have another check of args in `validate_op_between_ort_torch` which errors the validation if args are unsupported?

titaiwangms replied:
So I raise ValueError if it's unsupported types, and wrapped the whole `wrap_fx_args_as_torch_args` with try except ValueError to write the error message into SARIF.

BowenBao commented on torch/onnx/_internal/fx/fx_exporter.py:
```diff
@@ -49,10 +49,9 @@ def export_fx_to_onnx(
         # The pass added nodes with static shape into original node metadata:
-        # node.meta["node_with_static_shape"]: torch.fx.Node
-        # TODO(titaiwang): refactor the pass to stop relying on Transformer.transform()
+        # node.meta["static_shape"]: torch.fx.Node
```
update type annotation?

titaiwangms replied:
Done

BowenBao commented on torch/onnx/_internal/fx/passes/shape_inference.py:
```diff
@@ -19,13 +19,9 @@ class ShapeInferenceWithFakeTensor(_pass.Transform):
         # NOTE(titaiwang): Usually fx graph should have all the node meta value we need,
         # so we don't have to run FakeTensorProp to fill in node meta values. However, this
-        # can be used to validate op-level debugging when we only have symbolic shapes in
+        # is used to validate op-level debugging when we only have symbolic shapes in
```
```suggestion
        # is used to fill in static shapes, which are needed by op-level debugging, when we only have symbolic shapes in
```

titaiwangms replied:
Done

titaiwangms commented on torch/onnx/_internal/fx/op_validation.py:
```diff
@@ -195,14 +195,13 @@ def _recursive_wrap_args(
     for arg in complete_args:
         if isinstance(arg, torch.fx.Node):
-            fake_tensor = arg.meta["val"]
+            fake_tensor = arg.meta["static_shape"]
```
I can use .get("static_shape", None), and we only have to deal with it as one of the elif case.

titaiwangms committed:
Update on "[ONNX] Refactor ShapeInferenceWithFakeTensor to fill  metavalue into the original gm"


From https://github.com/pytorch/pytorch/pull/97494#discussion_r1160068456, the passes should modify gm inplace, but before this PR, `ShapeInferenceWithFakeTensor` utilizes Transform.transform() to make a copy of the gm, and rely on the assumption that the topological order of two graphs should be the same. This PR addresses the issue by saving another metavalue `static_shape` into gm for op_level_debug, instead of overwriting `val`.

cc BowenBao 

[ghstack-poisoned]

titaiwangms commented on torch/onnx/_internal/fx/op_validation.py:
```diff
@@ -195,14 +195,13 @@ def _recursive_wrap_args(
-                # TODO(titaiwang): Could dtype be inside a fx.Node?
                 wrapped_args.append(fake_tensor)
             else:
                 warnings.warn(
```
Done

titaiwangms committed:
Update on "[ONNX] Refactor ShapeInferenceWithFakeTensor to fill  metavalue into the original gm"


From https://github.com/pytorch/pytorch/pull/97494#discussion_r1160068456, the passes should modify gm inplace, but before this PR, `ShapeInferenceWithFakeTensor` utilizes Transform.transform() to make a copy of the gm, and rely on the assumption that the topological order of two graphs should be the same. This PR addresses the issue by saving another metavalue `static_shape` into gm for op_level_debug, instead of overwriting `val`.

cc BowenBao 

[ghstack-poisoned]

BowenBao commented on torch/onnx/_internal/fx/passes/shape_inference.py:
```diff
@@ -19,14 +19,10 @@ class ShapeInferenceWithFakeTensor(_pass.Transform):
+        # when we only have symbolic shapes in graph
 
         # Use this FakeTensorMode to
         # 1. convert nn.Parameter's in nn.Module to FakeTensor
```
FYI, https://github.com/pytorch/pytorch/pull/97186 has an update here. It shouldn't matter for this PR, but just in case.

titaiwangms replied:
Thanks!

titaiwangms committed:
Update on "[ONNX] Refactor ShapeInferenceWithFakeTensor to fill  metavalue into the original gm"


From https://github.com/pytorch/pytorch/pull/97494#discussion_r1160068456, the passes should modify gm inplace, but before this PR, `ShapeInferenceWithFakeTensor` utilizes Transform.transform() to make a copy of the gm, and rely on the assumption that the topological order of two graphs should be the same. This PR addresses the issue by saving another metavalue `static_shape` into gm for op_level_debug, instead of overwriting `val`.

cc BowenBao 

[ghstack-poisoned]

titaiwangms commented:
@pytorchbot merge
